src/tsd-metrics-factory.ts

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2016 Groupon.com
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import os = require("os");
+import path = require("path");
+
+import log4js = require("log4js");
+
+import events = require("events");
+import tsdDef = require("tsdDef");
+import tsdMetrics = require("./tsd-metrics");
+import options = require("./tsd-default-options");
+import log4jsSink = require("./sinks/tsd-query-log-sink");
+import consoleSink = require("./sinks/tsd-console-sink");
+import warningSink = require("./sinks/tsd-warning-sink");
+
+import TsdMetrics = tsdMetrics.TsdMetrics;
+import Options = options.Options;
+
+/**
+ * Factory pattern for creating and configuring instances of Metrics.
+ *
+ * @class
+ * @author Matthew Hayter (mhayter at groupon dot com)
+ */
+export class TsdMetricsFactory implements tsdDef.MetricsFactory {
+    private _serviceName:string;
+    private _clusterName:string;
+    private _hostName:string;
+    private _sinks:tsdDef.Sink[];
+    private _logger:log4js.Logger;
+
+    public static buildInstance(options: {serviceName:string; clusterName:string; hostName?:string; sinks?:tsdDef.Sink[]; logger?:log4js.Logger}):TsdMetricsFactory {
+        var failures:string[] = [];
+        var serviceName = options.serviceName;
+        var clusterName = options.clusterName;
+        var hostName = options.hostName;
+        var sinks = options.sinks;
+        var logger = options.logger;
+
+        if (clusterName == null) {
+            failures.push("clusterName must be provided");
+        }
+        if (serviceName == null) {
+            failures.push("serviceName must be provided");
+        }
+        if (hostName == null) {
+            hostName = os.hostname();
```

Comment:
We need to use a supplier pattern here to avoid possible name resolution which may block for several seconds. There will be a PR for the java client in the next week.